java/adapter/orc/src/main/java/org/apache/arrow/adapter/orc/OrcStripeReader.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.adapter.orc;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.stream.Collectors;
+
+import org.apache.arrow.flatbuf.MessageHeader;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.vector.ipc.ArrowReader;
+import org.apache.arrow.vector.ipc.ReadChannel;
+import org.apache.arrow.vector.ipc.message.ArrowDictionaryBatch;
+import org.apache.arrow.vector.ipc.message.ArrowFieldNode;
+import org.apache.arrow.vector.ipc.message.ArrowRecordBatch;
+import org.apache.arrow.vector.ipc.message.MessageChannelReader;
+import org.apache.arrow.vector.ipc.message.MessageResult;
+import org.apache.arrow.vector.ipc.message.MessageSerializer;
+import org.apache.arrow.vector.types.pojo.Schema;
+import org.apache.arrow.vector.util.ByteArrayReadableSeekableByteChannel;
+
+import io.netty.buffer.ArrowBuf;
+
+/**
+ * Orc stripe that load data into ArrowRecordBatch.
+ */
+public class OrcStripeReader extends ArrowReader {
+  private final long id;
+  private MessageChannelReader schemaReader;
+
+  /**
+   * Construct a new instance.
+   * @param id id of the stripe reader instance, obtained by
+   *           calling nextStripeReader from OrcReaderJniWrapper
+   * @param allocator memory allocator for accounting.
+   */
+  OrcStripeReader(long id, BufferAllocator allocator) {
+    super(allocator);
+    this.id = id;
+  }
+
+  @Override
+  public boolean loadNextBatch() throws IOException {
+    OrcRecordBatch recordBatch = OrcStripeReaderJniWrapper.next(id);
+    if (recordBatch == null) {
+      return false;
+    }
+
+    ArrayList<ArrowBuf> buffers = new ArrayList<>();
+    for (OrcMemoryJniWrapper buffer : recordBatch.buffers) {
+      buffers.add(new ArrowBuf(
+              new OrcReferenceManager(buffer),
+              null,
+              (int)buffer.getSize(),
```

Comment:
it would be good to throw when we truncate memory size (at least have a debug mode that does this).  Maybe not for this CL

Reply:
Not sure I fully understand you here. Under what circumstances would we truncate the memory size?

Reply:
if getSize() returns a value greater than MAX_INT

Reply:
I don't think we should solve such an issue here. If an arrow buffer size could go beyond MAX_INT, it will also cause issue on IPC ser/deser.